simpl/threadlocal.py

```diff
@@ -0,0 +1,52 @@
+"""Threadlocal."""
+import collections
+import threading
+
+THREAD_STORE = threading.local()
+DEFAULT_NAMESPACE = 'call_context'
+
+
+class ThreadLocalDict(collections.MutableMapping):
+
+    """A dict whose data is local to the thread."""
+
+    def __init__(self, varname, *args, **kwargs):
+        self.varname = varname
+        self.args = args
+        self.kwargs = kwargs
+
+    def __repr__(self):
+
+        under = repr(self._get_local_dict())
+        return '<ThreadLocalDict %s>' % under
+
+    def _get_local_dict(self):
+        """Retrieve (or initialize) the thread-local data to use."""
+        local_var = getattr(THREAD_STORE, self.varname, None)
+        if not local_var:
+            local_var = dict(*self.args, **self.kwargs)
+            setattr(THREAD_STORE, self.varname, local_var)
+        return local_var
+
+    def __len__(self):
+        return len(self._get_local_dict())
+
+    def __iter__(self):
+        return iter(self._get_local_dict())
+
+    def __getitem__(self, key):
+        return self._get_local_dict()[key]
+
+    def __setitem__(self, key, value):
+        self._get_local_dict()[key] = value
+
+    def __delitem__(self, key):
+        self._get_local_dict().__delitem__(key)
+
+
```

Comment:
I'm thinking about how to simplify the usage of this function from outside simpl. I wonder if everything below this line (and the definition of `DEFAULT_NAMESPACE` above) should live outside this module.

```python
import uuid

import simpl.context

CONTEXT = simple.context.default()

def some_thing(context):
    """Do something with context.""""
    context['transactionId'] = uuid.uuid4().hex

some_thing(CONTEXT)
print context.default()
#>>> <ThreadLocalContext {'transactionId': 'badabadabadabadabada'}>

# Custom context
print simple.context.new('custom_context').setdefault('foo', 1)
#>>> <ThreadLocalContext {'foo': 1}>
```

Not sure yet.... thinking out loud :-)